### PR TITLE
Make dead bodies disposable

### LINKED ISSUE #2340: You cant add dead mobs to a disposal unit
GIF 20-Oct-20 6-27-26 PM

### PR DESCRIPTION
**Pull request by Git-Nivrak**
Fixes #2340

### CODE REVIEW

**Comment by ike709** [comment]
Add `Fixes #2340` to the PR desc so github autocloses it.

**Comment by Git-Nivrak** [comment]
Fixes #2340

File: Content.Server/GameObjects/Components/Disposal/DisposalUnitComponent.cs [hunk 1]
```diff
@@ -136,10 +138,20 @@ public bool CanInsert(IEntity entity)
...
-            if (!entity.TryGetComponent(out IPhysicsComponent? physics) ||
+            IDamageableComponent? damageState;
+            entity.TryGetComponent<IDamageableComponent>(out damageState);
+
+            if (!entity.TryGetComponent(out IPhysicsComponent? physics) || 
                 !physics.CanCollide)
             {
-                return false;
+                if (damageState != null) {
+                    if (damageState.CurrentState != DamageState.Dead) {
```
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Disposal/DisposalUnitComponent.cs:148]: Merge the TryGetComponent with the `damageState` declaration, and also put the `TryGetComponent` where the current null check is. Afterwards, you can put the last if after the `TryGetComponent` using an and.

**Comment by Git-Nivrak** [comment]
Like so?

**Comment by Zumorica** [comment]
Indeed

**Comment by Zumorica** [comment]
Also CC @DrSmugleaf since this changes disposals

**Comment by DrSmugleaf** [comment]
> Also CC @DrSmugleaf since this changes disposals

I'm not a huge fan of special casing it for dead bodies in this way but it does fix the problem.
The CanCollide check was added after disabling collisions for entities in disposals to fix being teleported into space when going under a wall, I'm not sure if it would be easier to just remove that check entirely.

**Comment by Git-Nivrak** [comment]
I agree, I didn't like doing it either but dead bodies being uncollidable makes so many problems that require this special casing

**Comment by Git-Nivrak** [comment]
Also why do you need the check if you disable collisions anyways?

**Comment by Git-Nivrak** [comment]
Oops

**Comment by DrSmugleaf** [comment]
> Also why do you need the check if you disable collisions anyways?

The thought is that if you cant collide you shouldnt be able to be forced into a disposals tube, that of course breaks when its being disabled for dead bodies. Then again, dead bodies not being able to collide at all also means that you can pull move them through walls which is probably not desireable.

**Comment by Git-Nivrak** [comment]
Would it be that bad if it was on?

**Comment by DrSmugleaf** [comment]
> Would it be that bad if it was on?

Collision for dead bodies? Yes, unless their collisions layers or masks were changed since you would get blocked by dead bodies.
I think removing the CanCollide check from disposal units would be the saner option for now.

**Comment by Git-Nivrak** [comment]
I just tested, you don't get blocked by them

**Review by DrSmugleaf — APPROVED** [review]

**Pull request merged**